### ISSUE TITLE
Automatically look for and access connected code

### ISSUE THREAD

**Issue by jaykay9999** [microsoft/vscode-copilot-release#61]
### Personal Review

I honestly do not see the difference of using the chat service in vscode, or just copy and paste everything to ChatGPT on my web browser, infact, copying to ChatGPT-4 is much better. 

### Suggestion

What i suggest is that GitHub Copilot gets more access to the project code when asked about a specific snippet.

For example : 

1. I copy a code snippet that calls a function called "sort()"

2. Copilot automatically searches for the term in project files to find where that function is created and automatically adds it to the prompt (some prompt engineering is required here).

3. The task can be done again if the function "sort()" calls another function

4. Copilot stops at step 3 or 4 (which what I personally do when using ChatGPT to code apps, you can give the user to choose the number)

### Challenges

The chat history can easily exceed to allowed number of tokens, in this scenario, there are two possible solutions :

1- make Copilot look effectively only for the right code that might help it solve the problem.

2- take a huge chunk of the project code and Copilot automatically summarizes it to not lose context and understand the big image of the project. a 10 lines merge_sort python function is simply reduced to the text : "A merge sort function that takes a list input and return a list" using ChatGPT capabilities

**Comment by dbsxdbsx** [comment]
> I honestly do not see the difference of using the chat service in vscode, or just copy and paste everything to ChatGPT on my web browser, infact, copying to ChatGPT-4 is much better.

Just several months ago, I do agree.  But now since I can ask questions related to the project opened in vscode, it is better than chatgpt for a better understanding of the project I am dealing with. 

But still, I want to bump this issue for that... the current copilot chatbot can't update data. When I ask what is the latest version of Rust, it said (v1.55, now it is already v.170)--- This is even worse for flutter since the version changes lots of coding behavior.  

Thus, I hope web browsing could be added in next version, just like that in Edge.

**Comment by jaykay9999** [comment]
that is true, one big issue is that programming labguage documentation changes real quick overtime. 

gpt-4 can browse the web and it is actually amazing, i tried simply asking it to analyze tensorflow code on github for example and propose a template for me to add an optimizer that I will work on soon as a contribution and it did amazing. if github copilot has that ability it would be amazing. 

But there is always that issue of limited tokens, to what extent can it append text (meta prompts) to the chat. the only way to do that right from my perspective is that it summarizes information before adding it to the prompt as as a meta prompt.

I already suggested that as a feature to a library called auto-gpt ( literally uses chatgpt to code everything for you while youre just watching 😂 through terminal commands) 
and they opened a pull request for it and merged it.

I can refer to that issue here if they are thinking of implementing such logic.

**Comment by dbsxdbsx** [comment]
> But there is always that issue of limited tokens

Yes, from the [video](https://www.youtube.com/watch?v=CwAzIpc4AnA&t=2200s)(29:20),  now it is quite sure that due to limited tokens, currently, copilot can't gain all data of a project, the larger the less possible. But I think finally it would gain whole data one day, just like Claude-instant-100k.

**Comment by torsknod2** [comment]
Take care about the context automatically added by GitHub CoPilot.
Sometimes it makes sense to add things manually (I know inconvenient). 
But don't add too much, because this could overwhelm it.

Where I see a hard limitation right now is when working with binary dependencies, because even if available open source, it will not find them/ the right version and neither the documentation and use it as a context.
At least that is my impression.

**Comment by anthonykim1** [comment]
Im guessing this should be better now.
Try with latest copilot chat extension + [Insiders build](https://code.visualstudio.com/insiders)? Let us know if not.
